server/pbenchinacan/run-pbench-in-a-can

```diff
@@ -1,24 +1,74 @@
 #!/bin/bash
 
+# The default container registry to use, and the organization name in that
+# registry.
 export PB_CONTAINER_REG=${PB_CONTAINER_REG:-"images.paas.redhat.com"}
 export PB_ORG_NAME=${PB_ORG_NAME:-"pbench"}
 
+# Typically, the CI builds the Pbench Server container image and pushes it to
+# the container registry as part of the CI pipeline.  However, for development,
+# we need to build a local Pbench Server RPM and image from the checked-out
+# sources and run the functional tests against it.  But, we don't want to
+# require pushing that resulting image to the container registry in order to
+# pull it back to run it in the pod below.  To enable this, we use a separate
+# environment variable just for the Pbench Server.
+export PB_SERVER_CONTAINER_REG=${PB_SERVER_CONTAINER_REG:-${PB_CONTAINER_REG}}
+
+# Default pull policy for all images is to always try to pull the image (NOTE:
+# this does not mean it will download the image, but it will check to see if the
+# exact copy is available locally).
+export PB_COMMON_IMAGE_PULL_POLICY=${PB_COMMON_IMAGE_PULL_POLICY:-"Always"}
```

Comment:
I don't think this comment is correct.  I believe that `always` means, "pull a fresh copy every time it is run."  I believe that the option of "pull a fresh copy if the upstream version is not available locally" is `newer`:

From the `man` page:
```
    --pull=policy
       Pull image policy. The default is missing.

              • always: Always pull the image and throw an error if the pull fails.

              • missing: Pull the image only if it could not be found in the local
                containers storage.  Throw an error if no image could be found and
                the pull fails.

              • never: Never pull the image but use the one from the local containers
                storage.  Throw an error if no image could be found.

              • newer: Pull if the image on the registry is newer than the one in the
                local containers storage.  An image is considered to be newer when
                the digests are different.  Comparing  the time stamps is prone to
                errors.  Pull errors are suppressed if a local image was found.
```
So, I think we probably want `newer` here.

Reply:
> > How we name the images is less important than how we manage the pull policy. If we use a pull policy of `newer` (or `never`) instead of `always` (and, if we always name the images to match the remote repository), then I believe that we will be able to build and run repeatedly without anything having to be pulled (once the remote images are resident locally), and pushing the Pbench Server image is unnecessary in the build/run/repeat local scenario.
> 
> Kubernetes vs podman not withstanding (see https://kubernetes.io/docs/concepts/containers/images/#image-pull-policy, and PR #3216), I'll try not building with `localhost` and see how that behaves.

I tried this out and it "works," but I am a bit uncomfortable with the overlapping name space.

I'd rather see locally built images be local.  Then the container build scripts don't need to know where to target a build.  In both cases, you want to use a image pull policy for kubernetes of "never" anyways, so I think the extra mechanism is worth it to avoid the potential of somebody trying to run a local build, but via a mistake the remote build is used and they deem the local changes successful.

Reply:
I don't think the namespaces are actually overlapping.

When you pull an image from a remote registry, it is stored locally using the remote name.

Symmetrically, in the `podman push` command the "destination" argument is optional -- if the image name already includes the remote registry, then Podman will push it without further input.

So, I think the namespace is more "shared" than overlapping.

And, if you really want to operate "locally", you can set `PB_CONTAINER_REG` to `localhost`, and there won't be any confusion.  (You might need to re-tag the remote images after you've pulled them, but that is a one-time, quick and easy, if manual, operation.)

So, I really don't think we need to take any explicit action based on the value of the selected registry.  We just need to know whether the caller wants to push (or avoid pulling) any of the images...and, I think that that is completely separate of the target registry.

Reply:
Since I tried it out I have posted the changes.